File: src/data/meetings.rs
use anyhow::Result;
use chrono::{DateTime, Local, NaiveDateTime, TimeZone};
use std::process::Command;

use super::config::MeetingsConfig;

#[derive(Debug, Clone)]
pub struct MeetingsPanel {
    pub next_meeting: Option<Meeting>,
    pub upcoming: Vec<Meeting>,
    pub status: MeetingStatus,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Meeting {
    pub title: String,
    pub start: DateTime<Local>,
    pub end: DateTime<Local>,
    pub calendar: String,
}

#[derive(Debug, Clone)]
pub enum MeetingStatus {
    Free { minutes_until: i64 },
    InMeeting { ends_in: i64 },
    Clear,
}

impl Default for MeetingsPanel {
    fn default() -> Self {
        Self {
            next_meeting: None,
            upcoming: Vec::new(),
            status: MeetingStatus::Clear,
        }
    }
}

pub fn fetch_meetings(config: &MeetingsConfig) -> Result<MeetingsPanel> {
    if !config.enabled {
        return Ok(MeetingsPanel::default());
    }

    let excluded_cals: Vec<&str> = config
        .calendars_exclude
        .iter()
        .map(|s| s.as_str())
        .collect();
    let excluded_list = excluded_cals
        .iter()
        .map(|s| format!("\"{}\"", s))
        .collect::<Vec<_>>()
        .join(", ");

    let script = format!(
        r#"
        tell application "Calendar"
            set now to current date
            set endTime to now + ({} * hours)
            set output to ""
            repeat with cal in calendars
                if name of cal is not in {{{excluded_list}}} then
                    try
                        set evts to (every event of cal whose start date ≥ now and start date ≤ endTime)
                        repeat with evt in evts
                            set output to output & (summary of evt) & "|||"
                            set output to output & (start date of evt as string) & "|||"
                            set output to output & (end date of evt as string) & "|||"
                            set output to output & (name of cal) & "
"
                        end repeat
                    end try
                end if
            end repeat
            return output
        end tell
        "#,
        config.hours_ahead,
    );

    let output = Command::new("osascript").arg("-e").arg(&script).output();

    match output {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            parse_meetings(&stdout, config)
        }
        Err(_) => Ok(MeetingsPanel::default()),
    }
}

fn parse_meetings(output: &str, config: &MeetingsConfig) -> Result<MeetingsPanel> {
    let now = Local::now();
    let mut meetings: Vec<Meeting> = Vec::new();

    for line in output.lines() {
        if line.trim().is_empty() {
            continue;
        }

        let parts: Vec<&str> = line.split("|||").collect();
        if parts.len() < 4 {
            continue;
        }

        let title = parts[0].trim().to_string();

        // Skip ignored patterns
        if config
            .ignore_patterns
            .iter()
            .any(|p| title.to_lowercase().contains(&p.to_lowercase()))
        {
            continue;
        }

        let start = parse_applescript_date(parts[1].trim());
        let end = parse_applescript_date(parts[2].trim());
        let calendar = parts[3].trim().to_string();

        if let (Some(start), Some(end)) = (start, end) {
            meetings.push(Meeting {
                title,
                start,
                end,
                calendar,
            });
        }
    }

    // Sort by start time
    meetings.sort_by(|a, b| a.start.cmp(&b.start));

    // Determine status
    let status = if let Some(first) = meetings.first() {
        if first.start <= now && first.end > now {
            // Currently in a meeting
            let ends_in = (first.end - now).num_minutes();
            MeetingStatus::InMeeting { ends_in }
        } else {
            // Free until next meeting
            let minutes_until = (first.start - now).num_minutes();
            MeetingStatus::Free { minutes_until }
        }
    } else {
        MeetingStatus::Clear
    };

    let next_meeting = meetings.first().cloned();
    let upcoming = if meetings.len() > 1 {
        meetings[1..].to_vec()
    } else {
        Vec::new()
    };

    Ok(MeetingsPanel {
        next_meeting,
        upcoming,
        status,
    })
}

fn parse_applescript_date(date_str: &str) -> Option<DateTime<Local>> {
    // AppleScript dates look like: "Thursday, December 12, 2024 at 2:00:00 PM"
    // or sometimes: "December 12, 2024 2:00:00 PM"

    // Try multiple formats
    let formats = [
        "%A, %B %d, %Y at %I:%M:%S %p",
        "%B %d, %Y at %I:%M:%S %p",
        "%A, %B %d, %Y %I:%M:%S %p",
        "%B %d, %Y %I:%M:%S %p",
        "%m/%d/%Y %I:%M:%S %p",
        "%m/%d/%y %I:%M:%S %p",
    ];

    for fmt in &formats {
        if let Ok(naive) = NaiveDateTime::parse_from_str(date_str, fmt) {
            return Local.from_local_datetime(&naive).single();
        }
    }

    None
}

File: internal/data/meetings.go
package data

import (
	"fmt"
	"os/exec"
	"runtime"
	"sort"
	"strings"
	"time"

	"github.com/mrf/dcc/internal/config"
)

// Meeting represents a calendar meeting
type Meeting struct {
	Title    string
	Start    time.Time
	End      time.Time
	Calendar string
}

// MeetingStatus represents the current meeting status
type MeetingStatus int

const (
	StatusFree MeetingStatus = iota
	StatusInMeeting
	StatusClear
)

// MeetingsPanel holds meeting data for display
type MeetingsPanel struct {
	NextMeeting   *Meeting
	Upcoming      []Meeting
	Status        MeetingStatus
	MinutesUntil  int64
	EndsIn        int64
	IsLoading     bool
	Unsupported   bool
}

// FetchMeetings retrieves meetings from Calendar.app (macOS only)
func FetchMeetings(cfg config.MeetingsConfig) MeetingsPanel {
	if !cfg.Enabled {
		return MeetingsPanel{Status: StatusClear}
	}

	// Only works on macOS
	if runtime.GOOS != "darwin" {
		return MeetingsPanel{Status: StatusClear, Unsupported: true}
	}

	script := buildAppleScript(cfg)
	cmd := exec.Command("osascript", "-e", script)
	output, err := cmd.Output()
	if err != nil {
		return MeetingsPanel{Status: StatusClear}
	}

	meetings := parseMeetings(string(output), cfg.IgnorePatterns)
	return buildMeetingsPanel(meetings)
}

func buildAppleScript(cfg config.MeetingsConfig) string {
	excludeList := formatAppleScriptList(cfg.CalendarsExclude)

	return fmt.Sprintf(`
tell application "Calendar"
    set now to current date
    set endTime to now + (%d * hours)
    set output to ""
    repeat with cal in calendars
        if name of cal is not in %s then
            try
                set evts to (every event of cal whose start date >= now and start date <= endTime)
                repeat with evt in evts
                    set output to output & (summary of evt) & "|||" & (start date of evt) & "|||" & (end date of evt) & "|||" & (name of cal) & "\n"
                end repeat
            end try
        end if
    end repeat
    return output
end tell
`, cfg.HoursAhead, excludeList)
}

func formatAppleScriptList(items []string) string {
	if len(items) == 0 {
		return "{}"
	}
	quoted := make([]string, len(items))
	for i, item := range items {
		quoted[i] = fmt.Sprintf(`"%s"`, item)
	}
	return "{" + strings.Join(quoted, ", ") + "}"
}

func parseMeetings(output string, ignorePatterns []string) []Meeting {
	var meetings []Meeting
	lines := strings.Split(strings.TrimSpace(output), "\n")

	for _, line := range lines {
		if line == "" {
			continue
		}

		parts := strings.Split(line, "|||")
		if len(parts) != 4 {
			continue
		}

		title := strings.TrimSpace(parts[0])

		// Check if title matches ignore patterns
		skip := false
		for _, pattern := range ignorePatterns {
			if strings.Contains(strings.ToLower(title), strings.ToLower(pattern)) {
				skip = true
				break
			}
		}
		if skip {
			continue
		}

		start := parseAppleScriptDate(strings.TrimSpace(parts[1]))
		end := parseAppleScriptDate(strings.TrimSpace(parts[2]))
		calendar := strings.TrimSpace(parts[3])

		if start.IsZero() || end.IsZero() {
			continue
		}

		meetings = append(meetings, Meeting{
			Title:    title,
			Start:    start,
			End:      end,
			Calendar: calendar,
		})
	}

	// Sort by start time
	sort.Slice(meetings, func(i, j int) bool {
		return meetings[i].Start.Before(meetings[j].Start)
	})

	return meetings
}

func parseAppleScriptDate(dateStr string) time.Time {
	// AppleScript returns dates in various formats depending on system locale
	// Common formats: "Friday, January 26, 2024 at 10:00:00 AM"
	//                 "January 26, 2024 at 10:00:00 AM"
	//                 "26 Jan 2024, 10:00 AM"
	formats := []string{
		"Monday, January 2, 2006 at 3:04:05 PM",
		"January 2, 2006 at 3:04:05 PM",
		"Monday, January 2, 2006 at 15:04:05",
		"January 2, 2006 at 15:04:05",
		"2 Jan 2006, 3:04 PM",
		"2 Jan 2006, 15:04",
		"Monday, 2 January 2006 at 3:04:05 PM",
		"2 January 2006 at 3:04:05 PM",
		"Monday, 2 January 2006 at 15:04:05",
		"2 January 2006 at 15:04:05",
	}

	for _, format := range formats {
		if t, err := time.ParseInLocation(format, dateStr, time.Local); err == nil {
			return t
		}
	}

	return time.Time{}
}

func buildMeetingsPanel(meetings []Meeting) MeetingsPanel {
	now := time.Now()
	panel := MeetingsPanel{Status: StatusClear}

	if len(meetings) == 0 {
		return panel
	}

	// Check if currently in a meeting
	for i, m := range meetings {
		if now.After(m.Start) && now.Before(m.End) {
			panel.Status = StatusInMeeting
			panel.NextMeeting = &meetings[i]
			panel.EndsIn = int64(m.End.Sub(now).Minutes())
			if i+1 < len(meetings) {
				panel.Upcoming = meetings[i+1:]
			}
			return panel
		}
	}

	// Find next meeting
	for i, m := range meetings {
		if m.Start.After(now) {
			panel.Status = StatusFree
			panel.NextMeeting = &meetings[i]
			panel.MinutesUntil = int64(m.Start.Sub(now).Minutes())
			if i+1 < len(meetings) {
				panel.Upcoming = meetings[i+1:]
			}
			return panel
		}
	}

	return panel
}
